producers/prometheus/prometheus.go

```diff
@@ -15,17 +15,261 @@
+	"github.com/dcos/dcos-go/store"
 	"github.com/dcos/dcos-metrics/producers"
+	prodHelpers "github.com/dcos/dcos-metrics/util/producers"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var (
+	illegalChars = regexp.MustCompile("\\W")
+	promLog      = log.WithFields(log.Fields{"producer": "prometheus"})
 )
 
-type promProducer struct{}
+// Config is a configuration for the Prom producer's behaviour
+type Config struct {
+	Port        int `yaml:"port"`
+	CacheExpiry time.Duration
+}
+
+// promProducer implements both producers.MetricsProducer and
+// prometheus.Producer
+type promProducer struct {
+	config             Config
+	store              store.Store
+	metricsChan        chan producers.MetricsMessage
+	janitorRunInterval time.Duration
+}
 
-func New() (producers.MetricsProducer, chan producers.MetricsMessage) {
-	p := promProducer{}
-	c := make(chan producers.MetricsMessage)
-	return &p, c
+// New returns a prometheus producer and a channel for passing in metrics
+func New(cfg Config) (producers.MetricsProducer, chan producers.MetricsMessage) {
+	p := promProducer{
+		config:             cfg,
+		store:              store.New(),
+		metricsChan:        make(chan producers.MetricsMessage),
+		janitorRunInterval: 60 * time.Second,
+	}
+	return &p, p.metricsChan
 }
 
 func (p *promProducer) Run() error {
+	promLog.Info("Starting Prom producer garbage collection service")
+	go p.janitor()
+
+	// The below is borrowed wholesale from the http producer.
+	// TODO (philipnrmn): rewrite this section to avoid use of the store
+	go func() {
+		promLog.Debug("Prom producer listening for incoming messages on metricsChan")
+		for {
+			// read messages off the channel,
+			// and give them a unique name in the store
+			message := <-p.metricsChan
+
+			var name string
+			switch message.Name {
+			case producers.NodeMetricPrefix:
+				name = joinMetricName(message.Name, message.Dimensions.MesosID)
+
+			case producers.ContainerMetricPrefix:
+				name = joinMetricName(message.Name, message.Dimensions.ContainerID)
+
+			case producers.AppMetricPrefix:
+				name = joinMetricName(message.Name, message.Dimensions.ContainerID)
+			}
+
+			for _, d := range message.Datapoints {
+				p.writeObjectToStore(d, message, name)
+			}
+		}
+	}()
+
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(p)
+
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.HandlerFor(
+		registry, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}))
+
+	server := &http.Server{
+		Addr:    net.JoinHostPort("localhost", strconv.Itoa(p.config.Port)),
+		Handler: mux,
+	}
+
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			if err != http.ErrServerClosed {
+				promLog.Printf("E! Error creating prometheus metric endpoint, err: %s\n",
+					err.Error())
+			}
+		}
+	}()
 	return nil
 }
+
+// Describe passes all the available stat descriptions to Prometheus; we
```

Comment:
Why do we need to use a dummy metric to implement `Describe`? Some more doco here explaining that would be helpful.